Fix image sizes and srcset for intermediate images

mikeyarce opened the pull request:
## Description
This PR uses a new filter available in WP 5.7 to properly pass image dimensions for the image `srcset` and `sizes`.
This needs to be tested and merged in coordination with the WP 5.7 release (March 9 is the current tentative release date).

New filter added in WP 5.7:  https://core.trac.wordpress.org/browser/trunk/src/wp-includes/media.php?rev=50136#L1643

How the fix works:
- First we hook into `wp_img_tag_add_width_and_height_attr` which is early enough that our image source still has the query arguments available so we know what the image size is intended to be
- We save that image size data and then pass it into `wp_image_src_get_dimensions` where we set image dimensions
- If the image is a full size image, we set the default dimensions
- If the image is not full size, we calculate the image size based on the given width and the calculated aspect ratio.

## Changelog Description

#### Bug Fix: Fixed an issue where images were not properly sized when using the Block Editor and intermediate image sizes.

Previously, all images added through the Block Editor that were assigned intermediate image sizes ( e.g. Thumbnail ) would appear as full size images.

## Checklist

Please make sure the items below have been covered before requesting a review:

- [x] This change works and has been tested locally (or has an appropriate fallback).
- [x] This change works and has been tested on a Go sandbox.
- [ ] This change has relevant unit tests (if applicable).
- [ ] This change has relevant documentation additions / updates (if applicable).
- [x] I've created a changelog description that aligns with the provided examples. 

## Steps to Test
1. Patch your `wp-includes/media.php` file to include the changes to `wp_image_src_get_dimensions` found here: https://core.trac.wordpress.org/browser/trunk/src/wp-includes/media.php?rev=50136#L1601
2. Apply this code to your sandbox
3. Make sure your test site has Gutenberg Enabled
4. Create a new Post
5. Select the Image Block, add an image
6. From the Block Properties, select an intermediate image size like Thumbnail
7. Add more image blocks for Full Size, Medium, Large, etc..
8. Verify that the `srcset` and `sizes` attributes output are correct.

Before the change:
![before-patch-srcset](https://user-images.githubusercontent.com/3220162/109569248-fa6f0a00-7a9c-11eb-9788-fc2e3f700387.jpg)

After the change:
![after-patch-srcset](https://user-images.githubusercontent.com/3220162/109569279-09ee5300-7a9d-11eb-8824-2e3461a3d2bf.jpg)

mikeyarce committed:
Fix image sizes and srcset for intermediate images

Use a new filter available in WP 5.7 to properly pass image dimensions for the image srcset.

GaryJones commented:
Will those who are still on WP versions before 5.7, continue to see the same full size image for each named image size (i.e. the Before)?

GaryJones commented on a8c-files.php:
```diff
@@ -1156,6 +1156,59 @@ function a8c_files_maybe_inject_image_sizes( $data, $attachment_id ) {
+ * @param string $context       Additional context about how the function was called or where the img tag is.
+ * @param int    $attachment_id The image attachment ID.
+ */
+function a8c_files_maybe_inject_dimensions( $value, $image, $context, $attachment_id ) {
```
Any reason not to use type declarations on this parameters?

mikeyarce replied:
Good question!   We're not really using most of these since we're just returning `true` for the value in this filter and we're not returning any of them either, so I wasn't sure if that would be helpful in this case.  Would you suggest still adding them?  I also wasn't sure if I should pass them all even if we're not using them?

GaryJones replied:
If you're not using them, then there is no need to support them in the function signature, unless a param to the right is being used.

However, since you need `$attachment_id`, then they all need to be passed.

The non-first types should be known, as they are from core code, but the first type may have been filtered at an earlier priority into a different type, so although it's not being used here, the advantage of using types here is that it becomes an extra check in case some other code is doing something wrong. It's also good for IDEs and static analysis tools, like Psalm or Phpstan

Adding the return type also has no drawbacks.

GaryJones commented on a8c-files.php:
```diff
@@ -1156,6 +1156,59 @@ function a8c_files_maybe_inject_image_sizes( $data, $attachment_id ) {
+		return $dimensions;
+
+	}, 10, 4 );
+    return true;
```
Indented with 4 spaces instead of a tab character.

GaryJones commented on a8c-files.php:
```diff
@@ -1156,6 +1156,59 @@ function a8c_files_maybe_inject_image_sizes( $data, $attachment_id ) {
+	$vipgo_img_src = preg_match( '/src="([^"]+)"/', $image, $match_src ) ? $match_src[1] : '';
+
+	/** This filter is documented in wp-admin/includes/media.php */
+	add_filter( 'wp_image_src_get_dimensions', function( $dimensions, $image_src, $image_meta, $attachment_id ) use ( $vipgo_img_src ){
```
This closure is like 90% of the parent function.

Consider extracting it a named function, so that it can then be properly tested.

GaryJones commented on a8c-files.php:
```diff
@@ -1156,6 +1156,59 @@ function a8c_files_maybe_inject_image_sizes( $data, $attachment_id ) {
+				(int) $image_meta['height'],
+		);
+		// If it's not Full Size, let's use our calculated Aspect Ratio to output proper dimensions according to the size
+		} else if ( 
```
```suggestion
		} elseif ( 
```

mikeyarce replied:
I will make the change but I'm curious as to why.  From the PHP docs it looks like `else if` and `elseif` are the same when using curly brackets but not the same when using colons.  Is it just a stylistic choice in this case?

GaryJones replied:
@mikeyarce It's a WPCS choice: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/#use-elseif-not-else-if plus, `elseif` is then a unique keyword, should one need to find instances of it.

GaryJones commented on a8c-files.php:
```diff
@@ -1156,6 +1156,59 @@ function a8c_files_maybe_inject_image_sizes( $data, $attachment_id ) {
+				(int) $image_meta['width'],
+				(int) $image_meta['height'],
+		);
+		// If it's not Full Size, let's use our calculated Aspect Ratio to output proper dimensions according to the size
```
```suggestion
		// If it's not Full Size, let's use our calculated Aspect Ratio to output proper dimensions according to the size.
```

GaryJones commented on a8c-files.php:
```diff
@@ -1156,6 +1156,59 @@ function a8c_files_maybe_inject_image_sizes( $data, $attachment_id ) {
+				$dimensions = array(
+					(int) $query_args['w'],
+					(int) $query_args['h'],
+			);
```
```suggestion
				);
```

mikeyarce committed:
Indent with tab instead of spaces

mikeyarce committed:
Fix spacing and comment punctuation

mikeyarce committed:
Change `else if` to `elseif`

mikeyarce committed:
Move dimension functionality into separate function

mikeyarce commented:
Thanks for your review @GaryJones!  I've addressed the issues and separated out the dimension logic into another function.

> Will those who are still on WP versions before 5.7, continue to see the same full size image for each named image size (i.e. the Before)?

I tested this and for versions before 5.7, site's will continue to see the same `before` full size image for each named image.

rinatkhaziev committed:
Merge branch 'master' into fix-image-size-srcset

rinatkhaziev commented:
We're investigating other avenues, in the meantime, we'll keep it open but with the appropriate status.

mchanDev commented:
Hi @rinatkhaziev 

Is there a work around for this atm?

There is an open ticket here [#128167-z](https://wordpressvip.zendesk.com/agent/tickets/128167) that I would like to reply to
